backend/metadata_writer/src/metadata_writer.py

```diff
@@ -0,0 +1,756 @@
+        return execution_type
+
+
+def get_or_create_context_type(store, type_name, properties: dict = None) -> metadata_store_pb2.ContextType:
+    try:
+        context_type = store.get_context_type(type_name=type_name)
+        return context_type
+    except:
+        context_type = metadata_store_pb2.ContextType(
+            name=type_name,
+            properties=properties,
+        )
+        context_type.id = store.put_context_type(context_type) # Returns ID
+        return context_type
+
+
+def create_artifact_with_type(
+    store,
+    uri: str,
+    type_name: str,
+    properties: dict = None,
+    type_properties: dict = None,
+) -> metadata_store_pb2.Artifact:
+    artifact_type = get_or_create_artifact_type(
+        store=store,
+        type_name=type_name,
+        properties=type_properties,
+    )
+    artifact = metadata_store_pb2.Artifact(
+        uri=uri,
+        type_id=artifact_type.id,
+        properties=properties,
+    )
+    artifact.id = store.put_artifacts([artifact])[0]
+    return artifact
+
+
+def create_execution_with_type(
+    store,
+    type_name: str,
+    properties: dict = None,
+    type_properties: dict = None,
+) -> metadata_store_pb2.Execution:
+    execution_type = get_or_create_execution_type(
+        store=store,
+        type_name=type_name,
+        properties=type_properties,
+    )
+    execution = metadata_store_pb2.Execution(
+        type_id=execution_type.id,
+        properties=properties,
+    )
+    execution.id = store.put_executions([execution])[0]
+    return execution
+
+
+def create_context_with_type(
+    store,
+    context_name: str,
+    type_name: str,
+    properties: dict = None,
+    type_properties: dict = None,
+) -> metadata_store_pb2.Context:
+    # ! Context_name must be unique
+    context_type = get_or_create_context_type(
+        store=store,
+        type_name=type_name,
+        properties=type_properties,
+    )
+    context = metadata_store_pb2.Context(
+        name=context_name,
+        type_id=context_type.id,
+        properties=properties,
+    )
+    context.id = store.put_contexts([context])[0]
+    return context
+
+
+import functools
+@functools.lru_cache(maxsize=128)
+def get_context_by_name(
+    store,
+    context_name: str,
+) -> metadata_store_pb2.Context:
+    matching_contexts = [context for context in store.get_contexts() if context.name == context_name]
+    assert len(matching_contexts) <= 1
+    if len(matching_contexts) == 0:
+        raise ValueError('Context with name "{}" was not found'.format(context_name))
+    return matching_contexts[0]
+
+
+def get_or_create_context_with_type(
+    store,
+    context_name: str,
+    type_name: str,
+    properties: dict = None,
+    type_properties: dict = None,
+) -> metadata_store_pb2.Context:
+    try:
+        context = get_context_by_name(store, context_name)
+        context_types = store.get_context_types_by_id([context.type_id])
+        assert len(context_types) == 1
```

Comment:
> Can we only catch the specific type of error raised by MLMD here so the assertError can be escalate?

Do you mean outputting a meaningful error message?

Reply:
Yes that'll be great.

Specifically IMO we should error out when len(context_types) > 1. WDYT?

Reply:
OK. I understand what you mean now.
I thought that you wanted a different exception to be fired instead of AssertionError, but you were pointing out that this error is caught by try/catch. I'll think on how to change this part.

Reply:
Done.